### PR TITLE
Reset ALL connection settings for next loop iteration

### LINKED ISSUE #58787: Privilage escalation prompt timeout when working in a loop with different users
##### SUMMARY
Ansible reports password escalation prompt timeout when running a task with a loop, but only during 2nd iteration when using `ansible_connection=local`. The first iteration passes. See the attached playbook for a repro and details.

This is a change compared to older versions. I am not 100% sure but I think 2.7 did not have this behaviour, it started happening during last 1-2 months.

##### ISSUE TYPE
- Bug Report

##### COMPONENT NAME
become, privilege escalation

##### ANSIBLE VERSION

```
ansible 2.8.1
  config file = /etc/ansible/ansible.cfg
  configured module search path = ['/home/vespian/.ansible/plugins/modules', '/usr/share/ansible/plugins/modules']
  ansible python module location = /usr/lib/python3.7/site-packages/ansible
  executable location = /usr/bin/ansible
  python version = 3.7.3 (default, May 11 2019, 00:45:16) [GCC 8.3.1 20190223 (Red Hat 8.3.1-2)]
```

##### CONFIGURATION

```paste below
ANSIBLE_PIPELINING(/etc/ansible/ansible.cfg) = True
DEFAULT_FORKS(/etc/ansible/ansible.cfg) = 5
DEFAULT_GATHERING(/etc/ansible/ansible.cfg) = implicit
DEFAULT_JINJA2_EXTENSIONS(/etc/ansible/ansible.cfg) = jinja2.ext.do,jinja2.ext.i18n
DEFAULT_LOCAL_TMP(/etc/ansible/ansible.cfg) = /home/vespian/.ansible/tmp/ansible-local-17565m8xdgny3
DEFAULT_POLL_INTERVAL(/etc/ansible/ansible.cfg) = 15
DEFAULT_TRANSPORT(/etc/ansible/ansible.cfg) = ssh
DIFF_ALWAYS(/etc/ansible/ansible.cfg) = True
RETRY_FILES_ENABLED(/etc/ansible/ansible.cfg) = False
```
The problem manifests itself only when using `ansible_connection=local`, it works on remote hosts.

##### OS / ENVIRONMENT
x86_64, Fedora 29

##### STEPS TO REPRODUCE

```yaml
- name: Sample playbook
  hosts:
    - budrys
    - engel2
  tasks:
    - name: checkout plugins from upstream git repos
      command: whoami
      become: yes
      become_user: "{{ item[1] }}"
      with_nested:
        - - a
          - b
          - c
        - - vespian
          - root
          - kamila
```

##### EXPECTED RESULTS
The loop is executed without errors

##### ACTUAL RESULTS
The loop is unable to execute due to password prompt timeout error
```paste below
$ ap -l budrys -D -K /tmp/playbook.yml
BECOME password:

PLAY [Sample playbook] *******************************************************************************************************

TASK [Gathering Facts] *******************************************************************************************************
ok: [budrys]

TASK [checkout plugins from upstream git repos] ******************************************************************************
changed: [budrys] => (item=['a', 'vespian'])
changed: [budrys] => (item=['a', 'root'])
changed: [budrys] => (item=['a', 'kamila'])
fatal: [budrys]: FAILED! => {"msg": "timeout waiting for privilege escalation password prompt:\n"}

PLAY RECAP *******************************************************************************************************************
budrys                     : ok=1    changed=0    unreachable=0    failed=1    skipped=0    rescued=0    ignored=0
```
On a remote hosts, it works though (removed user `kamila` from the playbook as it is not present on the host`):
```
$ ap -l engel2 -D -K /tmp/playbook.yml
BECOME password:

PLAY [Sample playbook] **********************************************************************************************************************************

TASK [Gathering Facts] **********************************************************************************************************************************
ok: [engel2]

TASK [checkout plugins from upstream git repos] **********************************************************************************************************************************
changed: [engel2] => (item=['a', 'vespian'])
changed: [engel2] => (item=['a', 'root'])
changed: [engel2] => (item=['b', 'vespian'])
changed: [engel2] => (item=['b', 'root'])
changed: [engel2] => (item=['c', 'vespian'])
changed: [engel2] => (item=['c', 'root'])

PLAY RECAP **********************************************************************************************************************************
engel2                     : ok=2    changed=1    unreachable=0    failed=0    skipped=0    rescued=0    ignored=0

```

### PR DESCRIPTION
**Pull request by mkrizek**
##### SUMMARY

The original purpose of the change was to fix Become's
prompt not being reset for the next loop iteration which
this change fixes but also forces the connection/become/...
objects to be re-created which is simpler and more robust
than the original solution.

Co-authored-by: Brian Coca <bcoca@users.noreply.github.com>

Fixes #58787

##### ISSUE TYPE

- Bugfix Pull Request

##### COMPONENT NAME
`lib/ansible/executor/task_executor.py`

### CODE REVIEW

**Review by bcoca — APPROVED** [review]

**Comment by samdoran** [comment]
I re-ran tests and got the same failure again, which is not some we've seen before. It's probably a failure related to this PR.

**Review by bcoca — APPROVED** [review]

**Comment by nitzmahone** [comment]
Whoa, so we're getting rid of connection reuse in loops always? That's sure to piss off ... well, probably everyone that does things in loops.

Also kinda forces the issue on us knocking down the barriers to proper general connection reuse, but still...

**Comment by bcoca** [comment]
@nitzmahone  not really, connection reuse still happens with ssh as the persistence is handled outside of the plugin's life span. For networking the same is done via ansible-connection, Im still unsure about the repercussions on the winrm/psrp side, as it does look like we used to close the connection only after full loop iteration, still this can lead to wrong settings being used when they change per iteration.

I was thinking of moving this to the top of the loop and trying to detect changes before invalidating the existing connection.

If we used #69947 to store the actual connection linked to the hash, you have the reuse on 'match' built in.

**Comment by nitzmahone** [comment]
Something like that is probably what needs to happen anyway before we can do "real" connection reuse- as-is this seems like a pretty draconian measure for anyone that's not using ssh or networking.

**Comment by bcoca** [comment]
@nitzmahone most of the other methods don't currently have an effective reuse (most are forks to subprocs), but the winrm/psrp is something we should find a way around since they do seem to be impacted by this change.

**Review by bcoca — CHANGES_REQUESTED** [review]
need to find a way to not impact winrm/psrp reuse of connection (and x2 check no other plugins are impacted).

File: lib/ansible/plugins/become/sudo.py [hunk 1]
```diff
@@ -96,6 +96,8 @@ def build_become_command(self, cmd, shell):
             if flags:  # this could be simplified, but kept as is for now for backwards string matching
                 flags = flags.replace('-n', '')
             prompt = '-p "%s"' % (self.prompt)
+        else:
```
Comment (jimi-c) [inline lib/ansible/plugins/become/sudo.py:99]: This seems unrelated to the issue being fixed?

Comment (bcoca) [reply lib/ansible/plugins/become/sudo.py:99]: we also saw that if password presence changed between loop items the prompt would be incorrectly preserved.

File: lib/ansible/executor/task_executor.py [hunk 2]
```diff
@@ -414,19 +414,10 @@ def _run_loop(self, items):
...
-                    clear_plugins['become'] = self._connection.become._load_name
-
-                for plugin_type, plugin_name in iteritems(clear_plugins):
-                    for var in C.config.get_plugin_vars(plugin_type, plugin_name):
-                        if var in task_vars and var not in self._job_vars:
-                            del task_vars[var]
+            # needed for winrm/pspr before clearing the connection
+            self._connection.close()
+            # clear the connection for the next iteration that may have different options set
+            self._connection = None
```
Comment (bcoca) [inline lib/ansible/executor/task_executor.py:420]: giving it more thought, the fix here might be to just remove the old code 9it is really not doing what it is supposed to, task_vars is not updated anymore

then 'real fix' for the become issue is something like self.prompt = CLS.prompt in set_options before processing rest

**Review by bcoca — CHANGES_REQUESTED** [review]

File: lib/ansible/executor/task_executor.py (hunk 2 above)
Comment (mkrizek) [reply lib/ansible/executor/task_executor.py:420]: > giving it more thought, the fix here might be to just remove the old code 9it is really not doing what it is supposed to, task_vars is not updated anymore

We still need the original code to account for https://github.com/ansible/ansible/issues/58876:

```yaml
- hosts: test
  gather_facts: no
  tasks:
  - ping:
    remote_user: "{{ item }}"
    with_items:
      - centos
      - ec2-user
      - root
```

> then 'real fix' for the become issue is something like self.prompt = CLS.prompt in set_options before processing rest

So it sounds like we are back to https://github.com/ansible/ansible/pull/59764 although we need to override `set_options` in each become plugin since the default value of `prompt` differs per plugin.
